### PR TITLE
JavaScript 2018 (and beyond!)

### PR DESCRIPTION
**Pull request by roblarsen**
Some of the JavaScript questions are stale. I'm making an editorial pass to clean up those questions, removing stuff that's more 2008 than 2018.  I feel like some of these would stump people I'd be happy to hire. I am going to add a review to this PR to explain my reasoning for removing certain questions.

### CODE REVIEW

File: questions/javascript-questions.md [hunk 1]
```diff
@@ -1,21 +1,19 @@
 # JS Questions:
 
-* Explain event delegation
-* Explain how `this` works in JavaScript
-* Explain how prototypal inheritance works
-* What do you think of AMD vs CommonJS?
-* Explain why the following doesn't work as an IIFE: `function foo(){ }();`.
```
Comment (roblarsen) [inline questions/javascript-questions.md]: This seems like more of a trivia question than something relevant to most environment at present. in 2018. you could have five years of professional web development experience without having to use an IIFE.

File: questions/javascript-questions.md [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
+* Explain event delegation.
+* Explain how `this` works in JavaScript.
+  * Can you give an example of how `this` has changed in ES6?
+* Explain how prototypal inheritance works.
 * What's the difference between a variable that is: `null`, `undefined` or undeclared?
   * How would you go about checking for any of these states?
 * What is a closure, and how/why would you use one?
 * Can you describe the main difference between a `forEach` loop and a `.map()` loop and why you would pick one versus the other?
 * What's a typical use case for anonymous functions?
-* How do you organize your code? (module pattern, classical inheritance?)
```
Comment (roblarsen) [inline questions/javascript-questions.md]: This could be brought back, without the "module pattern, classical inheritance" part because even in robust frameworks and libraries people still have to make choices about organization.

File: questions/javascript-questions.md [hunk 3]
```diff
@@ -1,66 +1,54 @@
...
 * What's the difference between host objects and native objects?
-* Difference between: `function Person(){}`, `var person = Person()`, and `var person = new Person()`?
-* What's the difference between `.call` and `.apply`?
+* Explain the difference between: `function Person(){}`, `var person = Person()`, and `var person = new Person()`?
+* Explain the differences on the usage of `foo` between `function foo() {}` and `var foo = function() {}`
+* Can you explain what `.call` and `.apply` do? What's the notable difference between the two?
 * Explain `Function.prototype.bind`.
 * What's the difference between feature detection, feature inference, and using the UA string?
-* Explain Ajax in as much detail as possible.
```
Comment (roblarsen) [inline questions/javascript-questions.md]: Do people even talk about Ajax anymore? You could have a decade of experience and not really deal with "Ajax" in the way that this question implies.

File: questions/javascript-questions.md [hunk 4]
```diff
@@ -1,66 +1,54 @@
...
-* What are the advantages and disadvantages of using Ajax?
-* Explain how JSONP works (and how it's not really Ajax).
```
Comment (roblarsen) [inline questions/javascript-questions.md]: JSONP was a thing. It's no longer a thing.

File: questions/javascript-questions.md [hunk 5]
```diff
@@ -1,66 +1,54 @@
...
-* Have you ever used JavaScript templating?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: templates are ubiquitous

File: questions/javascript-questions.md [hunk 6]
```diff
@@ -1,66 +1,54 @@
...
-  * If so, what libraries have you used?
 * Explain "hoisting".
 * Describe event bubbling.
 * Describe event capturing.
 * What's the difference between an "attribute" and a "property"?
-* Why is extending built-in JavaScript objects not a good idea?
-* Difference between window load event and document DOMContentLoaded event?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: Party like it's 2005. Raise the roof for ppk and Dean Edwards.

File: questions/javascript-questions.md [hunk 7]
```diff
@@ -1,66 +1,54 @@
...
 * What is the difference between `==` and `===`?
 * Explain the same-origin policy with regards to JavaScript.
-* Make this work:
-```javascript
-duplicate([1,2,3,4,5]); // [1,2,3,4,5,1,2,3,4,5]
-```
 * Why is it called a Ternary operator, what does the word "Ternary" indicate?
-* What is `"use strict";`? what are the advantages and disadvantages to using it?
-* Create a for loop that iterates up to `100` while outputting **"fizz"** at multiples of `3`, **"buzz"** at multiples of `5` and **"fizzbuzz"** at multiples of `3` and `5`
-* Why is it, in general, a good idea to leave the global scope of a website as-is and never touch it?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: This could potentially return, if it's reworded. "Why is it dangerous to pollute the global namespace?" Probably not though. We've solved this problem.

File: questions/javascript-questions.md [hunk 8]
```diff
@@ -1,66 +1,54 @@
...
-* Why would you use something like the `load` event? Does this event have disadvantages? Do you know any alternatives, and why would you use those?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: two thousand zero zero five, party over, it's out of time.

File: questions/javascript-questions.md [hunk 9]
```diff
@@ -1,66 +1,54 @@
...
-* Explain what a single page app is and how to make one SEO-friendly.
```
Comment (roblarsen) [inline questions/javascript-questions.md]: We could add back in a question about SEO concerns of a SPA, but at this point failing on "what is a single page app" would be pretty bad. It implies that people are building apps that aren't single pages- for better or worse, they're not.

File: questions/javascript-questions.md [hunk 10]
```diff
@@ -1,66 +1,54 @@
...
-* What is the extent of your experience with Promises and/or their polyfills?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: 🤔

File: questions/javascript-questions.md [hunk 11]
```diff
@@ -1,66 +1,54 @@
...
-* What are the pros and cons of using Promises instead of callbacks?
```
Comment (roblarsen) [inline questions/javascript-questions.md]: Is this a conversation people are having now?

File: questions/javascript-questions.md (hunk 6 above)
Comment (roblarsen) [reply questions/javascript-questions.md]: Question: is http://dean.edwards.name really still running from his kitchen?

File: questions/javascript-questions.md (hunk 11 above)
Comment (vvscode) [reply questions/javascript-questions.md]: I can say that currently the question is promises vs async/await

File: questions/javascript-questions.md (hunk 3 above)
Comment (vvscode) [reply questions/javascript-questions.md]: For me this question is fine

File: questions/javascript-questions.md [hunk 12]
```diff
@@ -1,66 +1,54 @@
 # JS Questions:
 
-* Explain event delegation
-* Explain how `this` works in JavaScript
-* Explain how prototypal inheritance works
-* What do you think of AMD vs CommonJS?
-* Explain why the following doesn't work as an IIFE: `function foo(){ }();`.
-  * What needs to be changed to properly make it an IIFE?
+* Explain event delegation.
```
Comment (vvscode) [inline questions/javascript-questions.md]: Looks like we need grammar checker as test pipeline to avoid such changes ( I mean for new questions )

And single md-formatter

File: questions/javascript-questions.md (hunk 2 above)
Comment (vvscode) [reply questions/javascript-questions.md]: disagree about module pattern - it both about loading modules and abstractions for me. Probably we need rephrase it

File: questions/javascript-questions.md (hunk 5 above)
Comment (vvscode) [reply questions/javascript-questions.md]: now it might be about tagged-templates

File: questions/javascript-questions.md (hunk 3 above)
Comment (SenHeng) [reply questions/javascript-questions.md]: Maybe rephrase it to talk about the Fetch API instead, or ask about the differences between Fetch and Ajax?

File: questions/javascript-questions.md (hunk 4 above)
Comment (SenHeng) [reply questions/javascript-questions.md]: I believe JSONP is still commonly used by ad delivery networks.

Comment (roblarsen) [reply questions/javascript-questions.md]: It's a semi-thing? I don't think it's ubiquitous enough to mention here.

File: questions/javascript-questions.md (hunk 1 above)
Comment (allaev) [reply questions/javascript-questions.md]: I would suggest changing the 3rd question to: 
```suggestion
* What do you think of ESM vs CommonJS?
```
because AMD was pretty much replaced by ES modules (or ES6/ES2015 modules) and is rarely used nowadays.
Or alternatively:
```suggestion
* Where are ESM and CJS modules used?
```

File: questions/javascript-questions.md (hunk 10 above)
Comment (SenHeng) [reply questions/javascript-questions.md]: Considering the prevalence of IE11, this question is actually still quite relevant. Though I'd probably reword it to refer to polyfills in general.

**Comment by vvscode** [comment]
@roblarsen  Let's fix conflicts and merge ))) Good start for NY

**Comment by roblarsen** [comment]
@vvscode Sounds good. I was just thinking about this branch.

File: questions/javascript-questions.md [hunk 13]
```diff
@@ -1,66 +1,54 @@
...
+* Explain how `this` works in JavaScript.
+  * Can you give an example of how `this` has changed in ES6?
```
Comment (GMaiolo) [inline questions/javascript-questions.md]: I don't fancy the wording, I'd suggest something like 

> Can you give an example of how things involving `this` have changed with ES6?

Comment (vvscode) [reply questions/javascript-questions.md]: `this` is common term - why not?

Comment (GMaiolo) [reply questions/javascript-questions.md]: I mean that `this` didn't actually change in ES6 but things around `this` did. 
Just a simple semantic remark.

Comment (roblarsen) [reply questions/javascript-questions.md]: @GMaiolo I think it's a valid comment. I'll adjust the wording.

**Comment by roblarsen** [comment]
Going once... going twice... 

I'm ready to merge this.
